src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -123,12 +122,16 @@ protected override Payload ConstructHeadersPayload(GetHeadersPayload getHeadersP
                 return null;
             }
 
-            var headers = new ProvenHeadersPayload();
-            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            var provenHeadersPayload = new ProvenHeadersPayload();
+
+            ChainedHeader header = this.FindLastHeaderForPayload(fork, getHeadersPayload.HashStop);
+
+            for (int heightIndex = header.Height; heightIndex > fork.Height; heightIndex--)
             {
                 if (!(header.Header is ProvenBlockHeader provenBlockHeader))
                 {
                     this.logger.LogTrace("Invalid proven header, try loading it from the store.");
+
                     provenBlockHeader = this.provenBlockHeaderStore.GetAsync(header.Height).GetAwaiter().GetResult();
```

Comment:
ConstructHeadersPayload is called from async methods. 
I think maybe changing it to async and awaiting here instead of `.GetAwaiter().GetResult();` will be a good idea.

Reply:
Hmm we can but the base method in CMB doesn't do anything async?